modal/_utils/async_utils.py

```diff
@@ -463,3 +464,83 @@ def run_generator_sync(
         except BaseException as err:
             exc = err
     loop.close()
+
+
+@asynccontextmanager
+async def aclosing(gen: AsyncGenerator[T, None]) -> AsyncGenerator[AsyncGenerator[T, None], None]:
+    try:
+        yield gen
+    finally:
+        await gen.aclose()
+
+
+async def async_map(
+    input: AsyncIterator[T], async_mapper_func: Callable[[T], Awaitable[V]], concurrency: int
+) -> AsyncIterator[V]:
+    input_queue: asyncio.Queue[T] = asyncio.Queue(maxsize=concurrency)
+    results_queue: asyncio.Queue[V] = asyncio.Queue()
+    end_sentinel = object(T)
```

Comment:
What does `object(T)` mean? T is a typevar right - that wouldn't typically be used at runtime? 🤔

Reply:
A clean alternative to using end sentinels would be to use `inputs_queue.task_done()` and `inputs_queue.join()` btw, e.g.
Then you set `input_queue.task_done()` whenever you have put a new item on the result queue and can:

```
async def complete_map():
     await producer_task
     await inputs_queue.join()

complete_map_task = asyncio.create_task(complete_map())
```

`asyncio.wait([complete_map_task, has_new_result], when=FIST_COMPLETED)` to see when you need to either yield a new result or have exhausted inputs